ego/cli/bundle.go

```diff
@@ -0,0 +1,240 @@
+
+// ErrFileIsAlreadyBundled is returned when the file is already bundled. Technically, the outer binary is not an enclave either, but this is a more specific error.
+var ErrFileIsAlreadyBundled = errors.New("file is already bundled")
+
+// ErrFileIsNotAnEGoBinary is returned when the enclave has not been signed yet with 'ego sign', which would create a non-functional bundle.
+var ErrFileHasNotBeenSignedYet = errors.New("enclave has not been signed yet with 'ego sign'")
+
+// Bundle bundles an enclave with the currently installed EGo runtime into a single executable.
+func (c *Cli) Bundle(filename string) error {
+	// Check if the file is an enclave or already bundled
+	if err := c.checkIfBundable(filename); err != nil {
+		return err
+	}
+
+	// Build runtime image
+	runtimePath, err := c.buildImage(filename)
+	if err != nil {
+		return err
+	}
+	defer c.fs.Remove(runtimePath)
+
+	// Prepare the bundle
+	bundlePath, err := c.prepareBundle(filename)
+	if err != nil {
+		return err
+	}
+
+	// Add the runtime image to the bundle
+	if err := addSectionToELF(bundlePath, runtimePath, ".ego.bundle"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// checkIfBundable checks if the file is an enclave or already bundled.
+func (c *Cli) checkIfBundable(filename string) error {
+	file, err := c.fs.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	// Open the ELF file and check if a section called ".oeinfo" exists
+	elf, err := elf.NewFile(file)
+	if err != nil {
+		return err
+	}
+	defer elf.Close()
+
+	// Check if the ELF file has a section called ".oeinfo"
+	var isEnclave bool
+	var isAlreadyBundled bool
+	var isGoBinary bool
+	for _, section := range elf.Sections {
+		switch section.Name {
+		case ".oeinfo":
+			isEnclave = true
+		case ".ego.bundle":
+			isAlreadyBundled = true
+		case ".go.buildinfo":
+			isGoBinary = true
+		}
+	}
+
+	if isAlreadyBundled {
+		return ErrFileIsAlreadyBundled
+	}
+
+	if !isEnclave {
+		return ErrFileIsNotAnEGoBinary
+	}
+
+	if !isGoBinary {
+		return ErrFileIsNotAnEGoBinary
+	}
+
+	// Check if the binary has been signed with `ego sign` yet by checking if the payload has been appended.
+	// We do not check here if it is actually valid. If it is not valid, the binary can still be bundled, but during launch the enclave will refuse to run.
+	payloadSize, _, _, err := getPayloadInformation(file)
+	if err != nil {
+		return err
+	}
+
+	if payloadSize == 0 {
+		return ErrFileHasNotBeenSignedYet
+	}
+
+	return nil
+}
+
+// buildImage builds a gzip-compressed tarball containing the EGo runtime.
+func (c *Cli) buildImage(enclaveFilename string) (string, error) {
+	// Copy the runtime image to the tarPreparedContent
+	egoHostPath := c.getEgoHostPath()
+	egoEnclavePath := filepath.Join(c.egoPath, "share", "ego-enclave")
+
+	// Create the tar file
+	tempFile, err := c.fs.TempFile("", "ego-runtime")
+	if err != nil {
+		return "", err
+	}
+	defer tempFile.Close()
```

Comment:
remove the file if the func returns an err. See https://github.com/edgelesssys/constellation/blob/29bb781bb9e08eee5a81251915bf98eea50a3461/coordinator/pubapi/coord.go#L53-L54 on how to do this.